# Cyclistic Share Case Study
![""](Cyclistic_Bike_Share.jpg)
## How Does a Bike-Share Navigate Speedy Success?
- ### **Scenario**
You are a junior data analyst working in the marketing analyst team at Cyclistic, a bike-share company in Chicago. The director of marketing believes the company’s future success depends on maximizing the number of annual memberships. Therefore, your team wants to understand how casual riders and annual members use Cyclistic bikes differently. From these insights, your team will design a new marketing strategy to convert casual riders into annual members. But first, Cyclistic executives must approve your recommendations, so they must be backed up with compelling data insights and professional data visualizations.

- ### **Characters and Teams**
    - Cyclistic: A bike-share program that features more than 5,800 bicycles and 600 docking stations. Cyclistic sets itself apart by also offering reclining bikes, hand tricycles, and cargo bikes, making bike-share more inclusive to people with disabilities and riders who can’t use a standard two-wheeled bike. The majority of riders opt for traditional bikes; about 8% of riders use the assistive options. Cyclistic users are more likely to ride for leisure, but about 30% use them to commute to work each day.

    - Lily Moreno: The director of marketing and your manager. Moreno is responsible for the development of campaigns and initiatives to promote the bike-share program. These may include email, social media, and other channels.
    
    - Cyclistic marketing analytics team: A team of data analysts who are responsible for collecting, analysing, and reporting data that helps guide Cyclistic marketing strategy. You joined this team six months ago and have been busy learning about Cyclistic’s mission and business goals — as well as how you, as a junior data analyst, can help Cyclistic achieve them.

    - Cyclistic executive team: The notoriously detail-oriented executive team will decide whether to approve the recommended marketing program.

- ### **About the company**
In 2016, Cyclistic launched a successful bike-share offering. Since then, the program has grown to a fleet of 5,824 bicycles that are geotracked and locked into a network of 692 stations across Chicago. The bikes can be unlocked from one station and returned to any other station in the system anytime.

Until now, Cyclistic’s marketing strategy relied on building general awareness and appealing to broad consumer segments. One approach that helped make these things possible was the flexibility of its pricing plans: single-ride passes, full-day passes,
and annual memberships. Customers who purchase single-ride or full-day passes are referred to as casual riders. Customers who purchase annual memberships are Cyclistic members.

Cyclistic’s finance analysts have concluded that annual members are much more profitable than casual riders. Although the pricing flexibility helps Cyclistic attract more customers, Moreno believes that maximizing the number of annual members will be key to future growth. Rather than creating a marketing campaign that targets all-new customers, Moreno believes there is a very good chance to convert casual riders into members. She notes that casual riders are already aware of the Cyclistic
program and have chosen Cyclistic for their mobility needs.

Moreno has set a clear goal: Design marketing strategies aimed at converting casual riders into annual members. In order to do that, however, the marketing analyst team needs to better understand how annual members and casual riders differ, why
casual riders would buy a membership, and how digital media could affect their marketing tactics. Moreno and her team are interested in analysing the Cyclistic historical bike trip data to identify trends.


We will go through 6 phases of data analysis process

## **Ask**
Three questions will guide the future marketing program:
1. How do annual members and casual riders use Cyclistic bikes differently?
1. Why would casual riders buy Cyclistic annual memberships?
1. How can Cyclistic use digital media to influence casual riders to become members?

Moreno has assigned you the first question to answer:
How do annual members and casual riders use Cyclistic bikes differently?


#### **What is the problem I am trying to solve?**

Right now, this company focuses on broad customer segments and build general awareness. However, for the future profitable growth, the company targets turning casual riders into annual members rather than creating a new marketing campaign targeting all-new-customers. 
-	Why do they launch a new marketing campaign to target the existing customers instead of all-new-customers.
    -	Finance analyst concluded that annual members are much more profitable than casual riders.
-	Why is annual riders more profitable?
    - Maybe targeting on all-new-customers could not increase the number of new users anymore, so targeting on existing customers is more likely to gain profits. 
-	Why is targeting on existing customers better?
    - They are probably living around this area and use this service for the daily commuting purpose. If they are around this area, there is more chance that they will change to annual members. 
-   Why are there more chance?
    - In the big city, it will be more convinient to use bicycles to move around. If each station is located at convinient places, more local casual riders want to switch the membership to annual members.

##### **How can my insights drive business decisions?**

My insight can help Cyclistic evaluate promoting a new campaign for casual riders can be more effective and profitable than promoting a new campaign for all-new-customers like how it was before.

##### **A business task:**

Find and Compare patterns in annual riders and casual riders to see if a new campaign turning casual riders into annual members works more profitably.

##### **Key stakeholders**
- Primary Stakeholders
    -	Lily Moreno, Cyclistic Executive Team
- Secondary Stakeholders
    -	Cyclistic, Cyclistic Marketing Analytics Team 



## **Prepare&Process**

The data is located at [Index of bucket “divvy-tripdata”](https://divvy-tripdata.s3.amazonaws.com/index.html), and it contains trip_data from 2013-2022. This data gathers information from all users about how different customer types are using Cyclistic bike (e.g., rider_id, bicycle type, start and end time, start and end station, start and end station_id, start and end location coordination, member status). The data is less likely to be biased because it represents a whole rider, usage time, locations via geotracking. This data is owned by The City of Chicago under [this license](https://ride.divvybikes.com/data-license-agreement) and collected by this company’s first hand. The problem is that sometimes this geotracking system does not work, so some columns have null values. For the historical patterns like increasing of annual members, we will use data from 2013-2022. For the recent trend, we will use data from 2020-2022. We will download all data from the above link and save them on local storage based on the collected year.

![""](folder_order.jpg)


Some data is totally separated into a quarter or a month, so I will concatenate datasets into one dataset based on a year such as “Divvy_Trips_2014.csv.”
Here is an example.

we will concatenate and organize 2014_data.

First, we will import glob to select all files with the specific name and pandas to organize and clean data.

In [3]:
import glob
import pandas as pd

Now, we will extract all data with the name "Divvy_Trips_2014" under the specific path.

In [13]:
files=glob.glob("Trip_Dataset\data_2014\Divvy_Trips_2014*")

We will concatenate all datasets in the one.

In [14]:
data=pd.DataFrame()
for path in files:
    dataset=pd.read_csv(path)
    data=pd.concat([data,dataset],axis=0)

Check if the total length of each data matches with the length of new data.

In [15]:
length=0
for path in files:
    length+=len(pd.read_csv(path))
print(length, len(data))

2454634 2454634


It is perfectly successed. Let's take a look at data structures.

In [16]:
print(data.count())

trip_id              2454634
starttime            2454634
stoptime             2454634
bikeid               2454634
tripduration         2454634
from_station_id      2454634
from_station_name    2454634
to_station_id        2454634
to_station_name      2454634
usertype             2454634
gender               1663354
birthyear            1663418
dtype: int64


In [15]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2454634 entries, 0 to 905698
Data columns (total 12 columns):
 #   Column             Dtype  
---  ------             -----  
 0   trip_id            int64  
 1   starttime          object 
 2   stoptime           object 
 3   bikeid             int64  
 4   tripduration       int64  
 5   from_station_id    int64  
 6   from_station_name  object 
 7   to_station_id      int64  
 8   to_station_name    object 
 9   usertype           object 
 10  gender             object 
 11  birthyear          float64
dtypes: float64(1), int64(5), object(6)
memory usage: 243.5+ MB
None


All columns, except "gender" and "birthyear", have 2454634 entries, which means no missing entries.

The missing type of missing data in gender and birthyear columns is considered as Missing Not at Random(MNAR) because cyclists probably did not fill in these fields with reasons. Therefore, we will just leave it for now and inspect later.

However, to check if the starttime column properly has data from 2014-01-01 to 2014-12-31, we will change the data type of starttime and stoptime and sort it by starttime.

In [17]:
data["starttime"]=pd.to_datetime(data["starttime"])
data["stoptime"]=pd.to_datetime(data["stoptime"])
data.sort_values(by='starttime', inplace=True)

In [18]:
display(data)

trip_id           starttime            stoptime  bikeid  tripduration  \
905698  1109420 2014-01-01 00:17:00 2014-01-01 00:42:00    2981          1466   
905697  1109421 2014-01-01 00:45:00 2014-01-01 00:55:00    2981           608   
905696  1109427 2014-01-01 01:12:00 2014-01-01 01:18:00    1818           346   
905695  1109431 2014-01-01 01:43:00 2014-01-01 01:53:00     348           650   
905694  1109432 2014-01-01 01:43:00 2014-01-01 01:53:00     823           652   
...         ...                 ...                 ...     ...           ...   
4       4413163 2014-12-31 23:44:00 2015-01-01 00:10:00     389          1570   
3       4413164 2014-12-31 23:46:00 2015-01-01 00:37:00    2563          3084   
2       4413165 2014-12-31 23:48:00 2015-01-01 00:38:00    1946          3002   
1       4413166 2014-12-31 23:50:00 2014-12-31 23:52:00    1153           161   
0       4413167 2014-12-31 23:54:00 2014-12-31 23:57:00    1880           193   

        from_station_id             from_station_name  to_station_id  \
905698               94       Clark St & Armitage Ave             69   
905697               69        Damen Ave & Pierce Ave            216   
905696              240  Sheridan Rd & Irving Park Rd            245   
905695              113     Bissell St & Armitage Ave             94   
905694              113     Bissell St & Armitage Ave             94   
...                 ...                           ...            ...   
4                    43  Michigan Ave & Washington St            163   
3                   168        Michigan Ave & 14th St            168   
2                   168        Michigan Ave & 14th St            168   
1                   264  Stetson Ave & South Water St             44   
0                   296        Broadway & Belmont Ave            334   

                     to_station_name    usertype gender  birthyear  
905698        Damen Ave & Pierce Ave    Customer    NaN        NaN  
905697  California Ave & Division St    Customer    NaN        NaN  
905696    Clarendon Ave & Junior Ter  Subscriber   Male     1961.0  
905695       Clark St & Armitage Ave  Subscriber   Male     1988.0  
905694       Clark St & Armitage Ave  Subscriber   Male     1988.0  
...                              ...         ...    ...        ...  
4           Damen Ave & Clybourn Ave  Subscriber   Male     1983.0  
3             Michigan Ave & 14th St  Subscriber   Male     1987.0  
2             Michigan Ave & 14th St    Customer    NaN        NaN  
1             State St & Randolph St  Subscriber   Male     1984.0  
0        Lake Shore Dr & Belmont Ave  Subscriber   Male     1989.0  

[2454634 rows x 12 columns]

As we can see above, starttaime starts from 2014-01-01 to 2014-12-31. This is perfectly sorted. We will save this to csv file under the same path.


In [18]:
data.to_csv("Trip_Dataset\data_2014\Divvy_Trips_2014.csv", index=False)

Let's check if the data is correctly stored.

In [131]:
dataset=pd.read_csv("Trip_Dataset\data_2014\Divvy_Trips_2014.csv")

In [21]:
display(dataset)

trip_id            starttime             stoptime  bikeid  \
0        1109420  2014-01-01 00:17:00  2014-01-01 00:42:00    2981   
1        1109421  2014-01-01 00:45:00  2014-01-01 00:55:00    2981   
2        1109427  2014-01-01 01:12:00  2014-01-01 01:18:00    1818   
3        1109431  2014-01-01 01:43:00  2014-01-01 01:53:00     348   
4        1109432  2014-01-01 01:43:00  2014-01-01 01:53:00     823   
...          ...                  ...                  ...     ...   
2454629  4413163  2014-12-31 23:44:00  2015-01-01 00:10:00     389   
2454630  4413164  2014-12-31 23:46:00  2015-01-01 00:37:00    2563   
2454631  4413165  2014-12-31 23:48:00  2015-01-01 00:38:00    1946   
2454632  4413166  2014-12-31 23:50:00  2014-12-31 23:52:00    1153   
2454633  4413167  2014-12-31 23:54:00  2014-12-31 23:57:00    1880   

         tripduration  from_station_id             from_station_name  \
0                1466               94       Clark St & Armitage Ave   
1                 608               69        Damen Ave & Pierce Ave   
2                 346              240  Sheridan Rd & Irving Park Rd   
3                 650              113     Bissell St & Armitage Ave   
4                 652              113     Bissell St & Armitage Ave   
...               ...              ...                           ...   
2454629          1570               43  Michigan Ave & Washington St   
2454630          3084              168        Michigan Ave & 14th St   
2454631          3002              168        Michigan Ave & 14th St   
2454632           161              264  Stetson Ave & South Water St   
2454633           193              296        Broadway & Belmont Ave   

         to_station_id               to_station_name    usertype gender  \
0                   69        Damen Ave & Pierce Ave    Customer    NaN   
1                  216  California Ave & Division St    Customer    NaN   
2                  245    Clarendon Ave & Junior Ter  Subscriber   Male   
3                   94       Clark St & Armitage Ave  Subscriber   Male   
4                   94       Clark St & Armitage Ave  Subscriber   Male   
...                ...                           ...         ...    ...   
2454629            163      Damen Ave & Clybourn Ave  Subscriber   Male   
2454630            168        Michigan Ave & 14th St  Subscriber   Male   
2454631            168        Michigan Ave & 14th St    Customer    NaN   
2454632             44        State St & Randolph St  Subscriber   Male   
2454633            334   Lake Shore Dr & Belmont Ave  Subscriber   Male   

         birthyear  
0              NaN  
1              NaN  
2           1961.0  
3           1988.0  
4           1988.0  
...            ...  
2454629     1983.0  
2454630     1987.0  
2454631        NaN  
2454632     1984.0  
2454633     1989.0  

[2454634 rows x 12 columns]

The data for 2014 is stored into one file corractly and sorted by starttime properly.

The preparation step for 2014 is done.

We will automate all this process with the function below and organize all data. Save all into "Concat_Data"

In [126]:
import os 

In [128]:
def data_preparation():
    year=2013
    while year < 2023:
        select_path="Trip_Dataset\data_"+str(year)+"\*"
        
        files=glob.glob(select_path)
        data_trips=pd.DataFrame()
        data_stations=pd.DataFrame()

        for file in files:

            if "trip" in file.split('\\')[2].lower():
                print(file + " contains trip")
                if file.split(".")[1]=="csv":
                    dataset=pd.read_csv(file)
                elif file.split('.')[1]=="xlsx":
                    dataset=pd.read_excel(file)
                data_trips=pd.concat([data_trips, dataset],axis=0)

            elif "station" in file.split('\\')[2].lower():
                print(file + " contains station")
                if file.split(".")[1]=="csv":
                    dataset=pd.read_csv(file)
                elif file.split('.')[1]=="xlsx":
                    dataset=pd.read_excel(file)
                data_stations=pd.concat([data_stations,dataset], axis=0)
        storing_trips_path=os.path.join("Concat_Data","Divvy_Trips_"+str(year))
        storing_stations_path=os.path.join('Concat_Data',"Divvy_Stations_"+str(year))
        if not data_trips.empty:
            if os.path.exists(storing_trips_path+".csv")==False:
                data_trips.to_csv(storing_trips_path+".csv", index=False)
                print(storing_trips_path+".csv is added")
        if not data_stations.empty:
            if os.path.exists(storing_stations_path+".csv")==False:
                data_stations.to_csv(storing_stations_path+".csv", index=False)
                print(storing_stations_path+".csv is added")
        year+=1
    return None

In [129]:
data_preparation()

Trip_Dataset\data_2013\Divvy_Stations_2013.csv contains station
Trip_Dataset\data_2013\Divvy_Trips_2013.csv contains trip


C:\Users\crane\AppData\Local\Temp\ipykernel_10104\4047910637.py:15: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  dataset=pd.read_csv(file)


Trip_Dataset\data_2014\Divvy_Stations_2014-Q1Q2.xlsx contains station
Trip_Dataset\data_2014\Divvy_Stations_2014-Q3Q4.csv contains station
Trip_Dataset\data_2014\Divvy_Trips_2014-Q3-07.csv contains trip
Trip_Dataset\data_2014\Divvy_Trips_2014-Q3-0809.csv contains trip
Trip_Dataset\data_2014\Divvy_Trips_2014-Q4.csv contains trip
Trip_Dataset\data_2014\Divvy_Trips_2014.csv contains trip
Trip_Dataset\data_2014\Divvy_Trips_2014_Q1Q2.csv contains trip
Trip_Dataset\data_2015\Divvy_Stations_2015.csv contains station
Trip_Dataset\data_2015\Divvy_Trips_2015-Q1.csv contains trip
Trip_Dataset\data_2015\Divvy_Trips_2015-Q2.csv contains trip
Trip_Dataset\data_2015\Divvy_Trips_2015_07.csv contains trip
Trip_Dataset\data_2015\Divvy_Trips_2015_08.csv contains trip
Trip_Dataset\data_2015\Divvy_Trips_2015_09.csv contains trip
Trip_Dataset\data_2015\Divvy_Trips_2015_Q4.csv contains trip
Trip_Dataset\data_2016\Divvy_Stations_2016_Q1Q2.csv contains station
Trip_Dataset\data_2016\Divvy_Stations_2016_Q3.csv 

Let's check if the data is correctly stored

In [132]:
dataset1=pd.read_csv("Concat_Data\Divvy_Trips_2014.csv")
print(len(dataset), len(dataset1))

2454634 2454634


Now we can see data is successfully concatenated. As we can see below, Trips dataset is from 2013-2022, whereas Stations dataset is from 2013-2017.

![""](after_concat.jpg)

After I run display(csv_file), I find some data has wrong columns when it is concatnated. I will fix them so that all could have been prepared correctly.

- Modifying coulmns of Trip_Dataset\data_2018\Divvy_Trips_2018_Q1.csv and Trip_Dataset\data_2019\Divvy_Trips_2019_Q2.csv corresponding to other data columns.

#### Trips Data
- All trips_data contains start time and end time, but columns are not unified. We will modify the column names to "starttime" and "stoptime".
- Also, "starttime" and "stoptime" are time, so we will change data type to datetime.
- Moreover, tripduration can be converted into int64 instead of object type if there is no missing data.
- From 2013 to 2019, there is no missing data except gender and birthyear column. As I mentioned before, this is Missing Not at Random(MNAR) because there could be reasons behind this missing data. We will leave missing values.
- From 2020 to 2022, station name, id, longtitude, and latitude columns from start and end station are missing. We cannot fill all of missing data, but we will not drop rows with missing data because each row still contains useful information. We will try to find the start station name and id from start longtitude and latitude and find the end station name and id from end longtitude and latitude.
- We will add a tripduration column to dataset from 2020 to 2022.
- From starttime and stoptime, we will find the days of the week of usage and a month of usage.

In [4]:
files = glob.glob("Concat_Data/*")
trips_data={}
stations_data={}
year=2013
for file in files:
    if "trip" in file.split("\\")[1].lower():
        print(file.split("\\")[1].lower()+ " contains trip")
        trips_data[year]=pd.read_csv(file)
        print(str(year)+": "+file+" is added to trips_data")
        year+=1
year=2013
for file in files:
    if "station" in file.split("\\")[1].lower():
        print(file.split("\\")[1].lower()+ " contains station")
        stations_data[year]=pd.read_csv(file)
        print(str(year)+": "+file+" is added to stations_data")
        year+=1
print(trips_data.keys())
print(stations_data.keys())

divvy_trips_2013.csv contains trip


C:\Users\crane\AppData\Local\Temp\ipykernel_31720\1765331982.py:8: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  trips_data[year]=pd.read_csv(file)


2013: Concat_Data\Divvy_Trips_2013.csv is added to trips_data
divvy_trips_2014.csv contains trip
2014: Concat_Data\Divvy_Trips_2014.csv is added to trips_data
divvy_trips_2015.csv contains trip
2015: Concat_Data\Divvy_Trips_2015.csv is added to trips_data
divvy_trips_2016.csv contains trip
2016: Concat_Data\Divvy_Trips_2016.csv is added to trips_data
divvy_trips_2017.csv contains trip
2017: Concat_Data\Divvy_Trips_2017.csv is added to trips_data
divvy_trips_2018.csv contains trip
2018: Concat_Data\Divvy_Trips_2018.csv is added to trips_data
divvy_trips_2019.csv contains trip
2019: Concat_Data\Divvy_Trips_2019.csv is added to trips_data
divvy_trips_2020.csv contains trip


C:\Users\crane\AppData\Local\Temp\ipykernel_31720\1765331982.py:8: DtypeWarning: Columns (5,7) have mixed types. Specify dtype option on import or set low_memory=False.
  trips_data[year]=pd.read_csv(file)


2020: Concat_Data\Divvy_Trips_2020.csv is added to trips_data
divvy_trips_2021.csv contains trip
2021: Concat_Data\Divvy_Trips_2021.csv is added to trips_data
divvy_trips_2022.csv contains trip
2022: Concat_Data\Divvy_Trips_2022.csv is added to trips_data
divvy_stations_2013.csv contains station
2013: Concat_Data\Divvy_Stations_2013.csv is added to stations_data
divvy_stations_2014.csv contains station
2014: Concat_Data\Divvy_Stations_2014.csv is added to stations_data
divvy_stations_2015.csv contains station
2015: Concat_Data\Divvy_Stations_2015.csv is added to stations_data
divvy_stations_2016.csv contains station
2016: Concat_Data\Divvy_Stations_2016.csv is added to stations_data
divvy_stations_2017.csv contains station
2017: Concat_Data\Divvy_Stations_2017.csv is added to stations_data
dict_keys([2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022])
dict_keys([2013, 2014, 2015, 2016, 2017])


I see some numbers are stored as String. For example, the number was stored like 2,703. I will replce "," with "", so we can strip it and convert it into numeric type. Then, we will convert it into int64.

In [82]:
import numpy as np
for key, value in trips_data.items():
    print(key)
    value.rename(columns={"start_time":"starttime","end_time":"stoptime"}, inplace=True)
    value.rename(columns={"started_at":"starttime","ended_at":"stoptime"}, inplace=True)
    value.starttime=pd.to_datetime(value.starttime)
    value.stoptime=pd.to_datetime(value.stoptime)
    if "tripduration" in value.columns:
        if value.tripduration.dtype!="int64":
            value.tripduration=value.tripduration.str.replace(",","")
            value.tripduration=pd.to_numeric(value.tripduration)
            value.tripduration=value.tripduration.astype(np.int64)
    print(value.info())
    print(value.count())

2013
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 759788 entries, 0 to 759787
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   trip_id            759788 non-null  int64         
 1   starttime          759788 non-null  datetime64[ns]
 2   stoptime           759788 non-null  datetime64[ns]
 3   bikeid             759788 non-null  int64         
 4   tripduration       759788 non-null  int64         
 5   from_station_id    759788 non-null  int64         
 6   from_station_name  759788 non-null  object        
 7   to_station_id      759788 non-null  int64         
 8   to_station_name    759788 non-null  object        
 9   usertype           759788 non-null  object        
 10  gender             403046 non-null  object        
 11  birthday           402909 non-null  float64       
dtypes: datetime64[ns](2), float64(1), int64(5), object(4)
memory usage: 69.6+ MB
None
trip_id      

Let's fill the missing station name from longtitude and latitude.
We do not know the specific coordination of each station for 2020-2022 data, so we will find the mean latitude and longtitude for each station. Then, we will assign the closest station name to the unknown value.

In [17]:
data_trip_2020=trips_data[2020]
data_trip_2021=trips_data[2021]
data_trip_2022=trips_data[2022]


trips_from_2020_to_2022=[data_trip_2020,data_trip_2021,data_trip_2022]
stations_from_2020_to_2022={}

In [19]:

year=2020
for i in range(len(trips_from_2020_to_2022)):
    data=trips_from_2020_to_2022[i]
    start_latitude=data.groupby(["start_station_name"]).start_lat.mean()
    start_longtitude=data.groupby(["start_station_name"]).start_lng.mean()
    end_latitude=data.groupby(["end_station_name"]).end_lat.mean()
    end_longtitude=data.groupby(["end_station_name"]).end_lng.mean()
    stations={}
    for station in start_latitude.index:
        if station not in stations.keys():
            lat=start_latitude.loc[station]
            lng=start_longtitude.loc[station]
        else:
            lat=(stations[station][0]+start_latitude.loc[station])/2
            lng=(stations[station][1]+start_longtitude.loc[station])/2
        stations[station]=[lat,lng]

    for station in end_latitude.index:
        if station not in stations.keys():
            lat=end_latitude.loc[station]
            lng=end_longtitude.loc[station]
        else:
            lat=(stations[station][0]+end_latitude.loc[station])/2
            lng=(stations[station][1]+end_longtitude.loc[station])/2
        stations[station]=[lat,lng]
    stations_from_2020_to_2022[year]=stations
    year+=1

We have two stations only in start_station column and two stations only in end_station column. But, now we have the average coordination of each station in stations_2020 dictionary. 


We will extract rows with NaN values on start_station_name first and drop rows without start_lng or start_lat because we cannot compute the closest station without these data. This coordination error will take lots of decimals, so the error function will just calculate the absolute values of margin errors from each station.
We extract them and save it to "missing_station_name" with the key of year, and this will help reducing computational time.

In [20]:
def start_closest_station(row, year):
    closest=row.start_station_name
    stations=stations_from_2020_to_2022[year]
    if closest is np.nan:
        row_lat=row.start_lat
        row_lng=row.start_lng
        min_error=100
        for station in stations:
            lat,lng=stations[station]
            error=abs(lat-row_lat)+abs(lng-row_lng)
            if error<min_error:
                closest=station
    return closest


def end_closest_station(row, year):
    closest=row.end_station_name
    stations=stations_from_2020_to_2022[year]
    if closest is np.nan:
        row_lat=row.end_lat
        row_lng=row.end_lng
        min_error=100
        for station in stations:
            lat,lng=stations[station]
            error=abs(lat-row_lat)+abs(lng-row_lng)
            if error<min_error:
                closest=station
    return closest

We apply functions for each row and find the closest start station and the closest end station

In [21]:
missing_station_name={}
year=2020
for data in trips_from_2020_to_2022:
    missing_data=data[data.start_station_name.isna()]
    missing_data.dropna(subset=["start_lat","start_lng"], inplace=True)
    missing_data.start_station_name=missing_data.apply(start_closest_station, axis=1, year=(year))
    missing_data.end_station_name=missing_data.apply(end_closest_station, axis=1, year=(year))
    missing_station_name[year]=missing_data
    year+=1

C:\Users\crane\AppData\Local\Temp\ipykernel_31720\1407412702.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_data.dropna(subset=["start_lat","start_lng"], inplace=True)
C:\Users\crane\AppData\Local\Temp\ipykernel_31720\1407412702.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_data.start_station_name=missing_data.apply(start_closest_station, axis=1, year=(year))
C:\Users\crane\AppData\Local\Temp\ipykernel_31720\1407412702.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_in

In [22]:
missing_station_name_dict={}
def extract_data_into_dict(row):
    missing_station_name_dict[row.ride_id]=[row.start_station_name, row.end_station_name]

missing_station_name[2020].apply(extract_data_into_dict, axis=1)


630367     None
642029     None
653657     None
670853     None
670856     None
           ... 
3105966    None
3105967    None
3105968    None
3105969    None
3105970    None
Length: 94656, dtype: object

Below, functions will help filling dataset corresponding to missing_station_name, and we will replace the dataset with newly filled dataset.
We will overwrite old datasets with new datasets.

In [24]:
def fill_NaN(row):
    if row.isnull().start_station_name:
        if row.notnull().start_lng:
            row.start_station_name, row.end_station_name=missing_station_name_dict[row.ride_id]
    return row

In [41]:
missing_station_name_dict={}
missing_station_name[2020].apply(extract_data_into_dict, axis=1)
data_2020=trips_from_2020_to_2022[0]
filled_Data=data_2020.apply(fill_NaN, axis=1)


In [42]:
data_2020=filled_Data

In [27]:
missing_station_name_dict={}
missing_station_name[2021].apply(extract_data_into_dict, axis=1)
data_2021=trips_from_2020_to_2022[1]
filled_Data=data_2021.apply(fill_NaN, axis=1)

In [30]:
data_2021=filled_Data

In [32]:
missing_station_name_dict={}
missing_station_name[2022].apply(extract_data_into_dict, axis=1)
data_2022=trips_from_2020_to_2022[2]
filled_Data=data_2022.apply(fill_NaN, axis=1)

In [36]:
data_2022=filled_Data

In [ ]:
data_2020.to_csv("Concat_Data\Divvy_Trips_2020.csv", index=False)
data_2021.to_csv("Concat_Data\Divvy_Trips_2021.csv", index=False)
data_2022.to_csv("Concat_Data\Divvy_Trips_2022.csv", index=False)

Now, we will make a new column "tripduration" on datasets from 2020 to 2022 because they do not have it.
As I mentioned before, there is no missing values on starttime and stoptime, so we can just substract them.
tripduration needs to be in seconds

In [53]:
def trip_duration(row):
    return row.stoptime-row.starttime

In [64]:
data_2020["tripduration"]=data_2020.apply(trip_duration, axis=1)
data_2020["tripduration"]=data_2020.tripduration.dt.total_seconds().astype(np.int64)

In [65]:
data_2021["tripduration"]=data_2021.apply(trip_duration, axis=1)
data_2021["tripduration"]=data_2021.tripduration.dt.total_seconds().astype(np.int64)

In [66]:
data_2022["tripduration"]=data_2022.apply(trip_duration, axis=1)
data_2022["tripduration"]=data_2022.tripduration.dt.total_seconds().astype(np.int64)

In [73]:
data_2020.to_csv("Concat_Data\Divvy_Trips_2020.csv", index=False)
data_2021.to_csv("Concat_Data\Divvy_Trips_2021.csv", index=False)
data_2022.to_csv("Concat_Data\Divvy_Trips_2022.csv", index=False)

Now, we will add new columns, month and day_of_week, for all datasets.
We have updated csv files, so we will update our trips_data too.

In [78]:
files = glob.glob("Concat_Data/*")
trips_data={}
year=2013
for file in files:
    if "trip" in file.split("\\")[1].lower():
        print(file.split("\\")[1].lower()+ " contains trip")
        trips_data[year]=pd.read_csv(file)
        print(str(year)+": "+file+" is added to trips_data")
        year+=1
print(trips_data.keys())

divvy_trips_2013.csv contains trip


C:\Users\crane\AppData\Local\Temp\ipykernel_31720\4242506183.py:7: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  trips_data[year]=pd.read_csv(file)


2013: Concat_Data\Divvy_Trips_2013.csv is added to trips_data
divvy_trips_2014.csv contains trip
2014: Concat_Data\Divvy_Trips_2014.csv is added to trips_data
divvy_trips_2015.csv contains trip
2015: Concat_Data\Divvy_Trips_2015.csv is added to trips_data
divvy_trips_2016.csv contains trip
2016: Concat_Data\Divvy_Trips_2016.csv is added to trips_data
divvy_trips_2017.csv contains trip
2017: Concat_Data\Divvy_Trips_2017.csv is added to trips_data
divvy_trips_2018.csv contains trip
2018: Concat_Data\Divvy_Trips_2018.csv is added to trips_data
divvy_trips_2019.csv contains trip
2019: Concat_Data\Divvy_Trips_2019.csv is added to trips_data
divvy_trips_2020.csv contains trip


C:\Users\crane\AppData\Local\Temp\ipykernel_31720\4242506183.py:7: DtypeWarning: Columns (5,7) have mixed types. Specify dtype option on import or set low_memory=False.
  trips_data[year]=pd.read_csv(file)


2020: Concat_Data\Divvy_Trips_2020.csv is added to trips_data
divvy_trips_2021.csv contains trip
2021: Concat_Data\Divvy_Trips_2021.csv is added to trips_data
divvy_trips_2022.csv contains trip
2022: Concat_Data\Divvy_Trips_2022.csv is added to trips_data
dict_keys([2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022])


In [88]:
for data in trips_data.values():
    data["month"]=data.starttime.dt.month
    data["dayofweek"]=data.starttime.dt.dayofweek

#### Station Data from 2013 to 2017

- station data from 2017 has one extra column, so we just drop it

In [77]:
for key, value in stations_data.items():
    print(key)
    print(value.info())
    print(value.count())

2013
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           300 non-null    int64  
 1   name         300 non-null    object 
 2   latitude     300 non-null    float64
 3   longitude    300 non-null    float64
 4   dpcapacity   300 non-null    int64  
 5   landmark     300 non-null    int64  
 6   online date  300 non-null    object 
dtypes: float64(2), int64(3), object(2)
memory usage: 16.5+ KB
None
id             300
name           300
latitude       300
longitude      300
dpcapacity     300
landmark       300
online date    300
dtype: int64
2014
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           600 non-null    int64  
 1   name         600 non-null    object 
 2   latitude     600 

We just save all dataset under "Cleaned_Data" directory, so we can use them anytime.

In [109]:
from operator import index


def cleaned_data_to_csv(trips_data, stations_data):
    for key, value in trips_data.items():
        value.to_csv("Cleaned_Data\Divvy_Trips_"+str(key)+".csv", index=False)
    for key, value in stations_data.items():
        value.to_csv("Cleaned_Data\Divvy_Stations_"+str(key)+".csv", index=False)

In [110]:
cleaned_data_to_csv(trips_data, stations_data)

## **Analyse & Share** 
Now, we start analyzing the cleaned_data to see what difference casual riders and annual members.

We use R for data analysis to create and save visualizations.

The code is in the other file.

First, we will examine monthly usages between casual riders(Customer) and annual members(Subscriber) depending on years.

<img src="Monthly_usage_2013.png" width=50%><img src="Monthly_usage_2014.png" width=50%>
<img src="Monthly_usage_2015.png" width=50%><img src="Monthly_usage_2016.png" width=50%>
<img src="Monthly_usage_2017.png" width=50%><img src="Monthly_usage_2018.png" width=50%>
<img src="Monthly_usage_2019.png" width=50%><img src="Monthly_usage_2020.png" width=50%>
<img src="Monthly_usage_2021.png" width=50%><img src="Monthly_usage_2022.png" width=50%>
<img src="Usage_based_on_year.png" width=80%>
Findings:
- The summertime is always the peak. It seems the purpose of usage is for vacations.
- From 2013 to 2015, we can see a steady increase in both usertype usages. However, the number of casual riders gets decreased after that. Similarly, the number of casual riders also get decreased in 2017.
- After 2020, we can see the company changed its marketing strategy which targeted a new customer. The result is a steep increase in casual riders and a steep decrease in annual members. However, in 2021, both of the usertypes have a huge increase. During the summertime, we even can see that more casual riders use this service than annual riders.

Next, we analyze usage on each day of the week.
<img src=Usage_of_Dayofweek.png width=80%>

Findings:
- As we can see, casual riders use the service on the weekend, while annual members use the service on the weekdays.
- Casual rider usage has a slight decrease from Saturday to Wednesday, and it starts to increase on Saturday. From Friday to Saturday, there is a huge jump.
- Annual rider usage increase gradually from Monday to Tuesday, and then it starts decreasing toward Sunday, especially from Friday to Saturday. 
- We can assume simply that casual riders use the service for weekend trips, whereas annual members use the service for commuting to schools or work.

2013

<img src=2013_start_map.png width=50%><img src=2013_end_map.png width=50%>

2014

<img src=2014_start_map.png width=50%><img src=2014_end_map.png width=50%>

2015

<img src=2015_start_map.png width=50%><img src=2015_end_map.png width=50%>

2016

<img src=2016_start_map.png width=50%><img src=2016_end_map.png width=50%>

2017

<img src=2017_start_map.png width=50%><img src=2017_end_map.png width=50%>

2018

<img src=2018_start_map.png width=50%><img src=2018_end_map.png width=50%>

2019

<img src=2019_start_map.png width=50%><img src=2019_end_map.png width=50%>

2020

<img src=2020_start_map.png width=50%><img src=2020_end_map.png width=50%>

2021

<img src=2021_start_map.png width=50%><img src=2021_end_map.png width=50%>

2022

<img src=2022_start_map.png width=50%><img src=2022_end_map.png width=50%>

Findings:
- Both usertype dock a bicycle at a station from the close station where they ride on a bicycle.
- Casual riders mainly ride on bicycles around the airport, the beach and the museum, whereas annual members ride on a bicycle around main stations, the business district and the university. 
- Although the five most frequently used start stations and end stations for both usertypes look very similar and close, stations used by annual members look dense. Annual members use stations spreadly.

Also, we will take a look at the length of riding time between two user types.

<img src="Compare_Tripduration.jpg" width=50%>

Findings:
- Casual riders trip in the much longer distance than annual riders do. 

## **Act**
What can be suggested based on the above observaions?

First, let's revisit the key business task: *How do casual riders and annual members use their rental bikes differently?*

According to the observation, I conclude:
- They both have the peak of usage from both usertypes during the summertime, but the peak of annual riders is slightly later.
This can indicate that some use the service as casual riders first, and then switch to annual members after they experience this.
- Annual members seem to rent bicycles for commuting purposes because of the weekday usage, while casual riders rent bicycles for sightseeing or travelling through Chicago on the weekends. 
- Annual members rent bicycles from and stop bicycles at bike stations around train stations or the university, whereas casual riders frequently use stations around the airport, beaches, parks, or museums. This also supports the usage assumption of both users stated before. 
- Casual riders travel much longer distance compared to annual riders.

Suggestions:
- During the summertime, we can introduce a discount campaign and reduce each ride fee for a casual rider to attract more new customers. Also, we can have a summer trial campaign to let customers experience the benefit of having the membership over casual riders. Attracting more new customers and letting them experience the benefits of the membership can successfully convert some into annual members.
- Since casual riders rent a bike on the weekends hugely, this company can provide another flexibility in the service, which is "the weekend membership." To make this look more attractive, we can raise the fee of each ride for casual riders so that those users would feel "maybe it is better to join in the weekend membership" and switch to the membership.
- Having advertisements around the airport about this service and the convenience of talking about how long it will take to each landmark by bicycle can increase the potential of those tourists renting a bicycle. On the other hand, the company put advertisements about how beneficial having the membership is over casual rides around universities and train stations.